stdlib/open-uri.rb

```
@@ -0,0 +1,515 @@
+#
+# Author:: Tanaka Akira <akr@m17n.org>
+
+module OpenURI
+  Options = {
+    :proxy => true,
+    :proxy_http_basic_authentication => true,
+    :progress_proc => true,
+    :content_length_proc => true,
+    :http_basic_authentication => true,
+    :read_timeout => true,
+    :open_timeout => true,
+    :ssl_ca_cert => nil,
+    :ssl_verify_mode => nil,
+    :ftp_active_mode => false,
+    :redirect => true,
+    :encoding => nil,
+  }
+
+  def OpenURI.check_options(options) # :nodoc:
+    options.each {|k, v|
+      next unless Symbol === k
+      unless Options.include? k
+        raise ArgumentError, "unrecognized option: #{k}"
+      end
+    }
+  end
+
+  def OpenURI.scan_open_optional_arguments(*rest) # :nodoc:
+    if !rest.empty? && (String === rest.first || Integer === rest.first)
+      mode = rest.shift
+      if !rest.empty? && Integer === rest.first
+        perm = rest.shift
+      end
+    end
+    return mode, perm, rest
+  end
+
+  def self.open_uri(name, *rest) # :nodoc:
+    io = open_loop(name, {})
+    io.rewind
+    if block_given?
+      begin
+        yield io
+      ensure
+        if io.respond_to? :close!
+          io.close! # Tempfile
+        else
+          io.close if !io.closed?
+        end
+      end
+    else
+      io
+    end
+  end
+
+  def self.open_loop(uri, options) # :nodoc:
+    req = request(uri)
+    # FIXME Handle exception and status code ?
+    data = `req.responseText`
+    buf = Buffer.new
+    buf << data
+    io = buf.io
+    # FIXME Generate a URI object from the uri String
+    #io.base_uri = uri
+    status = `req.status`
+    statusText = `req.statusText`
+    io.status = "#{status} #{statusText}"
+    io.meta_add_field('content-type', `req.getResponseHeader("Content-Type")`)
+    io
+  end
+
+  def self.request(uri)
+    %x{
+      var xhr = new XMLHttpRequest();
+      xhr.open('GET', uri, false);
+      xhr.send();
+      return xhr;
+    }
+  end
+
+  class HTTPError < StandardError
```

If the return code is not 200, 301, 302, 303 or 307, we should throw this exception. Also if something bad happens with XHR (for instance an invalid URI), we should also throw this exception.